src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs

```diff
@@ -353,7 +353,10 @@ public virtual void Write(string value)
 
             if (_useFastUtf8)
             {
-                if (value.Length <= 127 / 3)
+                // If this is a non-derived BinaryWriter, then we can bypass the Write7BitEncodedInt call.
+                // But when this is a derived instance, call must not bypass it for compatibility reasons
+                // as it calls the virtual Write(int) overload.
+                if (GetType() == typeof(BinaryWriter) && value.Length <= 127 / 3)
```

Comment:
@EgorBo How is runtime optimizing the logic that uses `GetType() == typeof(BinaryWriter)` pattern as of today? Will this be always replaced with a constant for each derived type? If not, should we cache the result of the check? Or at least change the order (perform the cheap length check first)

Reply:
It's a single pointer check (`*object == TypeTableOf(BinaryWriter)`) and equivalent to accessing a field. This pattern is used a lot currently.